### DISCUSSION TITLE
Auto Cull 2d text

### DISCUSSION THREAD

**Discussion by miketwenty1** [bevyengine/bevy#13367] [Q&A] [answered]
I'm trying to cull 2d text so my game doesn't lag, I'm spawning a massive amount of 2d text in my game, and I want an easy way to hide it, if it's not insight of the camera. 

Is there an easy way to do this without recreating the wheel? I see there is "ViewVisibility" https://docs.rs/bevy/latest/bevy/prelude/struct.ViewVisibility.html, but I'm unsure how to use it. 

Any pointers or examples would be great!

**Comment by rparrett** [comment]
I suspect that adding an `Aabb` component to the text entity would be enough to make culling work.

**Reply by miketwenty1** [reply]
@rparrett Would you be able to provide an example of adding a lightweight phony `Aabb` component to an entity/2d text?

```rust
#[derive(Component)]
pub struct CullingAabbComponent(pub Aabb);


let phony = Aabb {
    center: Vec3A::new(0.0, 0.0, 0.0),
    half_extents: Vec3A::new(0.0, -0.0, 0.0),
};

```
```rust
...

builder.spawn((
  Text2dBundle {
      // ...
      ..default()
  },
  CullingAabbComponent(phony),
  ));
  ...     
```

I tried adding this component to my text2d bundle and it didn't seem to make a difference.

**Reply by rparrett** [reply]
Sure. Here's a quick and dirty adaptation of our `many_sprites` example, but with `Text2d` instead and a system to insert `Aabb` when the computed text layout info changes.

https://gist.github.com/rparrett/9cd3b145def1855f1baf87822e6a5f7a

**Reply by miketwenty1** [reply]
It's unclear to me what you're doing, on every loop you are inserting a new Aabb into Text2d entities?

In terms of what it should be doing.. is this different than what I'm attempting where I'm including Aabb when spawning the Text2dBundle. Are you able to confirm that your method actually works? Do you see anything wrong with my approach of including it as a component when spawning? I'm not seeing any changes.

**Reply by rparrett** [reply, marked as answer]
> on every loop you are inserting a new Aabb into Text2d entitie

Only when `TextLayoutInfo` is updated.

> Are you able to confirm that your method actually works

Yes, you can test this for yourself by removing the `update_text_aabb` from the app and observing the logged output as well as what should be an obvious degradation in frame rate.

> Do you see anything wrong with my approach of including it as a component when spawning

Yes, the Bevy system that determines visibility based on `Aabb` is querying for an `Aabb`, so embedding the `Aabb` within your `CullingAabbComponent` won't work.

Adding the `Aabb` at spawn time will work fine, but I am adding it when `TextLayoutInfo` is updated so that I can give the `Aabb` correct dimensions.

**Reply by miketwenty1** [reply]
@rparrett this worked amazingly after just "`insert`"ing like you are doing after I spawn the text entity.  Thanks for the help! my game no longer lags!
